enhancements/20230830-managed-dhcp.md

```diff
@@ -0,0 +1,318 @@
+1. Create another cluster network called `provider-cp` (the cluster network consists of **only the control plane nodes**)
+1. Create a network config called `provider-cp-nc` for the cluster network `provider-cp` with only the control plane nodes and one of their NICs as the uplink
+1. Create a VM network called `priv-net-all` with cluster network `provider-all` selected. The VLAN ID should be set with a number which is not in use in the premises, say, `817`.
+1. Create another VM network called `priv-net-cp` with cluster network `provider-cp` selected. The VLAN ID should be set with a number which is not in use in the premises and the above VM network, say, `689`.
+1. Create a VM called `test-vm` with two NICs attaching to the `priv-net-all` and the `priv-net-cp` networks respectively. Also, the NIC's MAC address for `priv-net-all` is `02:d4:6f:a8:bc:ed` and the other for `priv-net-cp` is `24:ab:56:78:9c:01`.
+
+From now on, the steps will interact with the managed DHCP feature.
+
+1. Create an IPPool object `priv-net-all` with the following YAML content:
+
+    ```yaml
+    apiVersion: network.harvesterhci.io/v1alpha1
+    kind: IPPool
+    metadata:
+      name: priv-net-all
+      namespace: default
+    spec:
+      ipv4Config:
+        serverIP: 172.19.150.2
+        cidr: 172.19.150.0/28
+        pool:
+        start: 172.19.150.5
+        end: 172.19.150.14
+        leaseTime: 600
+      networkName: default/priv-net-all
+    ```
+
+2. Create another IPPool object `priv-net-cp` with the following YAML content:
+
+    ```yaml
+    apiVersion: network.harvesterhci.io/v1alpha1
+    kind: IPPool
+    metadata:
+      name: priv-net-cp
+      namespace: default
+    spec:
+      ipv4Config:
+        serverIP: 172.19.100.2
+        cidr: 172.19.100.0/28
+        pool:
+        start: 172.19.100.10
+        end: 172.19.100.14
+        leaseTime: 300
+      networkName: default/priv-net-cp
+    ```
+
+3. Create a VirtualMachineNetworkConfig object for the VM `test-vm-1`:
+
+    ```yaml
+    apiVersion: network.harvesterhci.io/v1alpha1
+    kind: VirtualMachineNetworkConfig
+    metadata:
+      name: test-vm-1
+      namespace: default 
+    spec:
+      vmName: test-vm-1
+      networkConfig:
+      - macAddress: 02:d4:6f:a8:bc:ed
+        networkName: default/priv-net-all
+        ipAddress: 172.19.150.9
+      - macAddress: 24:ab:56:78:9c:01
+        networkName: default/priv-net-cp
+    ```
+
+4. Check the following:
+
+    - The `default-priv-net-cp-agent` agent Pod should run on one of the control plane nodes
+    - The two IPPool objects' `.status` field should contain the correct allocation records and statistics
+    - The VirtualMachineNetworkConfig object's `.status` field should show the two NICs have IP addresses allocated
+    - The VM `test-vm-1` should have the correct IP addresses configured automatically via DHCP from the agents respectively
+
+#### Kubernetes Guest Clusters (Rancher Integration)
```

Comment:
do we need changes to the machine driver to pass additional info to VM for leveraging the managed dhcp feature?

Reply:
vm-dhcp-controller listens to VMI-related events and creates the corresponding vmnetcfg object; then, the IP address can be allocated from the pool by the vmnetcfg reconcile loop. This is only for the dynamic IP allocation scenario. As @ibrokethecloud suggested, we need other ways to support the static IP allocation if the user tends to configure the VM's IP address statically during the VM creation. This is tracked in #5052. Thank you!